// @privy-io/react-auth/dist/cjs/Layouts-D3TOcPhm.js
"use strict";var e=require("styled-components");const t=e.styled.div`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  justify-content: flex-end;
  margin-top: auto;
  gap: 16px;
  flex-grow: 100;
`,i=e.styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  flex-grow: 1;
  width: 100%;
`,o=e.styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
`,r=e.styled(i)`
  padding: 20px 0;
`,d=e.styled(i)`
  gap: 16px;
`,l=e.styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
`,n=e.styled.div`
  display: flex;
  flex-direction: column;
  gap: 8px;
`;e.styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  height: 100%;
`;const s=e.styled.div`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: flex-start;
  text-align: left;
  gap: 8px;
  padding: 16px;
  margin-top: 16px;
  margin-bottom: 16px;
  width: 100%;
  background: var(--privy-color-background-2);
  border-radius: var(--privy-border-radius-md);
  && h4 {
    color: var(--privy-color-foreground-3);
    font-size: 14px;
    text-decoration: underline;
    font-weight: medium;
  }
  && p {
    color: var(--privy-color-foreground-3);
    font-size: 14px;
  }
`,a=e.styled.div`
  height: 16px;
`,p=e.styled.div`
  height: 12px;
`;e.styled.div`
  position: relative;
`;const c=e.styled.div`
  height: ${e=>e.height??"12"}px;
`;e.styled.div`
  background-color: var(--privy-color-accent);
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 50%;
  border-color: white;
  border-width: 2px !important;
`,exports.BottomPusherContainer=t,exports.CenteredItem=i,exports.CenteredItemWithGap=d,exports.CenteredItemWithPadding=r,exports.ConfigurableSpacer=c,exports.FixedGappedContainer=n,exports.FlexContainer=l,exports.HorizontallyCenteredItem=o,exports.RefactorSpacerBottom=p,exports.RefactorSpacerTop=a,exports.StyledCalloutSection=s;
