// @privy-io/react-auth/dist/cjs/OpenLink-L5LZ-pEm.js
"use strict";var r=require("react/jsx-runtime"),e=require("react");let o=require("styled-components").styled.a`
  display: flex;
  align-items: center;
  gap: 6px;

  && {
    margin: 8px 2px;
    font-size: 14px;
    color: ${r=>r.justOpened?"var(--privy-color-foreground)":r.color||"var(--privy-color-foreground-3)"};
    font-weight: ${r=>r.justOpened?"medium":"normal"};
    transition: color 350ms ease;

    :focus,
    :active {
      background-color: transparent;
      border: none;
      outline: none;
      box-shadow: none;
    }

    :hover {
      color: ${r=>r.justOpened?"var(--privy-color-foreground)":"var(--privy-color-foreground-2)"};
    }

    :active {
      color: 'var(--privy-color-foreground)';
      font-weight: medium;
    }

    @media (max-width: 440px) {
      margin: 12px 2px;
    }
  }

  svg {
    width: 14px;
    height: 14px;
  }
`;exports.OpenLinkButton=t=>{let[n,i]=e.useState(!1);/*#__PURE__*/return r.jsx(o,{color:t.color,href:t.url,target:"_blank",rel:"noreferrer noopener",onClick:()=>{i(!0),setTimeout((()=>i(!1)),1500)},justOpened:n,children:t.text})};
