// @privy-io/react-auth/dist/cjs/styles-Csarwypj.js
"use strict";var t=require("styled-components"),e=require("./ModalHeader-C5O9gNUP.js"),n=require("./LinkPasskeyScreen-o4qdrAL3.js");const o=t.styled.div`
  display: flex;
  flex-direction: column;
  gap: 12px;
  padding-top: 24px;
  padding-bottom: 24px;
`,r=t.styled.div`
  width: 24px;
  height: 24px;
  display: flex;
  justify-content: center;
  align-items: center;

  svg {
    border-radius: var(--privy-border-radius-sm);
  }
`,i=t.styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: flex-start;
  gap: 8px;
`,s=t.styled.div`
  display: flex;
  align-items: center;
  gap: 4px;
  width: 100%;
  padding: 0 16px;
  border-width: 1px !important;
  border-radius: 12px;
  cursor: text;

  &:focus-within {
    border-color: var(--privy-color-accent);
  }
`;t.styled.div`
  font-size: 42px !important;
`;const d=t.styled.input`
  background-color: var(--privy-color-background);
  width: 100%;

  &:focus {
    outline: none !important;
    border: none !important;
    box-shadow: none !important;
  }

  && {
    font-size: 26px;
  }
`,a=t.styled(d)`
  && {
    font-size: 42px;
  }
`;t.styled.button`
  cursor: pointer;
  padding-left: 4px;
`;const p=t.styled.div`
  font-size: 18px;
`,l=t.styled.div`
  font-size: 12px;
  color: var(--privy-color-foreground-3);
  // we need this container to maintain a static height if there's no content
  height: 20px;
`;t.styled.div`
  display: flex;
  flex-direction: row;
  line-height: 22px;
  font-size: 16px;
  text-align: center;
  svg {
    margin-right: 6px;
    margin: auto;
  }
`,t.styled(n.LinkButton)`
  margin-top: 16px;
`;let c=t.keyframes`
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
`;t.styled(e.SoftCtaButton)`
  border-radius: var(--privy-border-radius-md) !important;
  animation: ${c} 0.3s ease-in-out;
`;const u=t.styled.div``,x=t.styled.a`
  && {
    color: var(--privy-color-accent);
  }

  cursor: pointer;
`;exports.FundingAmountInput=d,exports.FundingAmountInputLarge=a,exports.FundingCurrency=p,exports.FundingDollars=l,exports.FundingMethodContainer=o,exports.FundingQuantity=s,exports.FundingQuantityWrapper=i,exports.IconContainer=r,exports.InfoButtonLink=x,exports.InfoButtonText=u;
