// @privy-io/react-auth/dist/cjs/TodoList-Cii_KrAk.js
"use strict";var e=require("react/jsx-runtime"),r=require("lucide-react"),t=require("react"),i=require("styled-components");function o(e){return e&&"object"==typeof e&&"default"in e?e:{default:e}}var s=/*#__PURE__*/o(t);const n=i.styled.div`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  gap: 10px; /* 10px gap between items */
  padding-left: 8px; /* 8px indentation container */
`;i.styled.div`
  &&& {
    margin-left: 6px; /* Center the line under the checkbox (12px/2) */
    border-left: 2px solid var(--privy-color-foreground-4);
    height: 10px; /* 10px H padding between paragraphs */
    margin-top: 0;
    margin-bottom: 0;
  }
`;let a=i.styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 20px;
  height: 20px;
  border-radius: 50%;
  background-color: ${({$variant:e})=>{switch(e){case"success":return"var(--privy-color-success-bg, #EAFCEF)";case"error":return"var(--privy-color-error-bg, #FEE2E2)";default:return"var(--privy-color-background-2)"}}};
  flex-shrink: 0;
`,c=i.styled.div`
  display: flex;
  justify-content: flex-start;
  align-items: flex-start; /* Align all elements to the top */
  text-align: left;
  gap: 8px;

  && {
    a {
      color: var(--privy-color-accent);
    }
  }
`;exports.TodoItem=({children:t,variant:i="default",icon:o})=>{let n=()=>{switch(i){case"success":return"var(--privy-color-icon-success)";case"error":return"var(--privy-color-icon-error)";default:return"var(--privy-color-icon-muted)"}};/*#__PURE__*/return e.jsxs(c,{children:[/*#__PURE__*/e.jsx(a,{$variant:i,"data-variant":i,children:(()=>{if(o)
/*#__PURE__*/return s.default.isValidElement(o)?/*#__PURE__*/s.default.cloneElement(o,{stroke:n(),strokeWidth:2}):o;switch(i){case"success":default:/*#__PURE__*/return e.jsx(r.Check,{size:12,stroke:n(),strokeWidth:3});case"error":/*#__PURE__*/return e.jsx(r.X,{size:12,stroke:n(),strokeWidth:3})}})()}),t]})},exports.TodoList=n;
