// @privy-io/react-auth/dist/cjs/WalletLink-CZLtiI7E.js
"use strict";var e=require("react/jsx-runtime"),r=require("styled-components"),t=require("viem"),i=require("./ethers-XrYC9HG-.js"),s=require("./getFormattedUsdFromLamports-Mu2fqwL2.js"),n=require("./transaction-DjF5X-by.js"),l=require("./useActiveWallet-vlXeDPzp.js");let a=r.styled.span`
  font-size: 14px;
  line-height: 140%;
  display: flex;
  gap: 4px;
  align-items: center;
`,o=r.styled.span`
  font-size: 12px;
  line-height: 12px;
  color: var(--privy-color-foreground-3);
`,d=r.styled.span`
  font-size: 10px;
`,c=r.styled.span`
  display: flex;
  align-items: center;
`;function u(e,r){return`https://explorer.solana.com/account/${e}?chain=${r}`}let m=r.styled.a`
  &:hover {
    text-decoration: underline;
  }
`;exports.HeaderPriceDisplay=({weiQuantities:r,tokenPrice:t,tokenSymbol:s})=>{let n=i.sumWeiQuantities(r),l=t?i.getDollarsFromWei(n,t):void 0,o=i.getNativeCurrencyFromWei(n,s);/*#__PURE__*/return e.jsx(a,{children:l||o})},exports.PriceDisplay=({weiQuantities:r,tokenPrice:t,tokenSymbol:s})=>{let n=i.sumWeiQuantities(r),l=t?i.getDollarsFromWei(n,t):void 0,u=i.getNativeCurrencyFromWei(n,s);/*#__PURE__*/return e.jsx(a,{children:l?/*#__PURE__*/e.jsxs(e.Fragment,{children:[/*#__PURE__*/e.jsx(o,{children:"USD"}),"<$0.01"===l?
/*#__PURE__*/e.jsxs(c,{children:[/*#__PURE__*/e.jsx(d,{children:"<"}),"$0.01"]}):l]}):u})},exports.SolanaPriceDisplay=({quantities:r,tokenPrice:i,tokenSymbol:l="SOL",tokenDecimals:o=9})=>{let u=r.reduce(((e,r)=>e+r),0n),m=i&&"SOL"===l&&9===o?s.getFormattedUsdFromLamports(u,i):void 0,p="SOL"===l&&9===o?n.getNativeCurrencyFromLamports(u):`${t.formatUnits(u,o)} ${l}`;/*#__PURE__*/return e.jsx(a,{children:m?/*#__PURE__*/e.jsx(e.Fragment,{children:"<$0.01"===m?
/*#__PURE__*/e.jsxs(c,{children:[/*#__PURE__*/e.jsx(d,{children:"<"}),"$0.01"]}):m}):p})},exports.WalletLink=r=>/*#__PURE__*/e.jsx(m,{href:"ethereum"===r.chainType?i.getBlockExplorerUrlForAddress(r.chainId,r.walletAddress):u(r.walletAddress,r.chainId),target:"_blank",children:l.formatWalletAddress(r.walletAddress)});
