// @privy-io/api-types/resources/wallets/wallets.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Wallets = void 0;
const tslib_1 = require("../../internal/tslib.js");
const resource_1 = require("../../core/resource.js");
const BalanceAPI = tslib_1.__importStar(require("./balance.js"));
const balance_1 = require("./balance.js");
const TransactionsAPI = tslib_1.__importStar(require("./transactions.js"));
const transactions_1 = require("./transactions.js");
const pagination_1 = require("../../core/pagination.js");
const headers_1 = require("../../internal/headers.js");
const path_1 = require("../../internal/utils/path.js");
class Wallets extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.transactions = new TransactionsAPI.Transactions(this._client);
        this.balance = new BalanceAPI.Balance(this._client);
    }
    /**
     * Creates a new wallet on the requested chain and for the requested owner.
     *
     * @example
     * ```ts
     * const wallet = await client.wallets.create({
     *   chain_type: 'ethereum',
     * });
     * ```
     */
    create(params, options) {
        const { 'privy-idempotency-key': privyIdempotencyKey, ...body } = params;
        return this._client.post('/v1/wallets', {
            body,
            ...options,
            headers: (0, headers_1.buildHeaders)([
                { ...(privyIdempotencyKey != null ? { 'privy-idempotency-key': privyIdempotencyKey } : undefined) },
                options?.headers,
            ]),
        });
    }
    /**
     * Get all wallets in your app.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const wallet of client.wallets.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/v1/wallets', (pagination_1.Cursor), { query, ...options });
    }
    /**
     * Export a wallet's private key
     *
     * @example
     * ```ts
     * const response = await client.wallets._export('wallet_id', {
     *   encryption_type: 'HPKE',
     *   recipient_public_key:
     *     'BDAZLOIdTaPycEYkgG0MvCzbIKJLli/yWkAV5yCa9yOsZ4JsrLweA5MnP8YIiY4k/RRzC+APhhO+P+Hoz/rt7Go=',
     * });
     * ```
     */
    _export(walletID, params, options) {
        const { 'privy-authorization-signature': privyAuthorizationSignature, ...body } = params;
        return this._client.post((0, path_1.path) `/v1/wallets/${walletID}/export`, {
            body,
            ...options,
            headers: (0, headers_1.buildHeaders)([
                {
                    ...(privyAuthorizationSignature != null ?
                        { 'privy-authorization-signature': privyAuthorizationSignature }
                        : undefined),
                },
                options?.headers,
            ]),
        });
    }
    /**
     * Initialize a wallet import. Complete by submitting the import.
     *
     * @example
     * ```ts
     * const response = await client.wallets._initImport({
     *   address: 'address',
     *   chain_type: 'ethereum',
     *   encryption_type: 'HPKE',
     *   entropy_type: 'hd',
     *   index: 0,
     * });
     * ```
     */
    _initImport(body, options) {
        return this._client.post('/v1/wallets/import/init', { body, ...options });
    }
    /**
     * Sign a message with a wallet by wallet ID.
     *
     * @example
     * ```ts
     * const response = await client.wallets._rawSign(
     *   'wallet_id',
     *   {
     *     params: {
     *       hash: '0x0775aeed9c9ce6e0fbc4db25c5e4e6368029651c905c286f813126a09025a21e',
     *     },
     *   },
     * );
     * ```
     */
    _rawSign(walletID, params, options) {
        const { 'privy-authorization-signature': privyAuthorizationSignature, 'privy-idempotency-key': privyIdempotencyKey, ...body } = params;
        return this._client.post((0, path_1.path) `/v1/wallets/${walletID}/raw_sign`, {
            body,
            ...options,
            headers: (0, headers_1.buildHeaders)([
                {
                    ...(privyAuthorizationSignature != null ?
                        { 'privy-authorization-signature': privyAuthorizationSignature }
                        : undefined),
                    ...(privyIdempotencyKey != null ? { 'privy-idempotency-key': privyIdempotencyKey } : undefined),
                },
                options?.headers,
            ]),
        });
    }
    /**
     * Sign a message or transaction with a wallet by wallet ID.
     *
     * @example
     * ```ts
     * const response = await client.wallets._rpc('wallet_id', {
     *   method: 'personal_sign',
     *   params: { encoding: 'utf-8', message: 'message' },
     * });
     * ```
     */
    _rpc(walletID, params, options) {
        const { 'privy-authorization-signature': privyAuthorizationSignature, 'privy-idempotency-key': privyIdempotencyKey, ...body } = params;
        return this._client.post((0, path_1.path) `/v1/wallets/${walletID}/rpc`, {
            body,
            ...options,
            headers: (0, headers_1.buildHeaders)([
                {
                    ...(privyAuthorizationSignature != null ?
                        { 'privy-authorization-signature': privyAuthorizationSignature }
                        : undefined),
                    ...(privyIdempotencyKey != null ? { 'privy-idempotency-key': privyIdempotencyKey } : undefined),
                },
                options?.headers,
            ]),
        });
    }
    /**
     * Submit a wallet import request.
     *
     * @example
     * ```ts
     * const wallet = await client.wallets._submitImport({
     *   wallet: {
     *     address: '0xF1DBff66C993EE895C8cb176c30b07A559d76496',
     *     chain_type: 'ethereum',
     *     ciphertext:
     *       'PRoRXygG+YYSDBXjCopNYZmx8Z6nvdl1D0lpePTYZdZI2VGfK+LkFt+GlEJqdoi9',
     *     encapsulated_key:
     *       'BOhR6xITDt5THJawHHJKrKdI9CBr2M/SDWzZZAaOW4gCMsSpC65U007WyKiwuuOVAo1BNm4YgcBBROuMmyIZXZk=',
     *     encryption_type: 'HPKE',
     *     entropy_type: 'private-key',
     *   },
     * });
     * ```
     */
    _submitImport(body, options) {
        return this._client.post('/v1/wallets/import/submit', { body, ...options });
    }
    /**
     * Update a wallet's policies or authorization key configuration.
     *
     * @example
     * ```ts
     * const wallet = await client.wallets._update('wallet_id');
     * ```
     */
    _update(walletID, params, options) {
        const { 'privy-authorization-signature': privyAuthorizationSignature, ...body } = params;
        return this._client.patch((0, path_1.path) `/v1/wallets/${walletID}`, {
            body,
            ...options,
            headers: (0, headers_1.buildHeaders)([
                {
                    ...(privyAuthorizationSignature != null ?
                        { 'privy-authorization-signature': privyAuthorizationSignature }
                        : undefined),
                },
                options?.headers,
            ]),
        });
    }
    /**
     * Obtain a session key to enable wallet access.
     *
     * @example
     * ```ts
     * const response = await client.wallets.authenticateWithJwt({
     *   user_jwt:
     *     'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWUsImlhdCI6MTUxNjIzOTAyMn0.KMUFsIDTnFmyG3nMiGM6H9FNFUROf3wh7SmqJp-QV30',
     * });
     * ```
     */
    authenticateWithJwt(body, options) {
        return this._client.post('/v1/wallets/authenticate', { body, ...options });
    }
    /**
     * Create wallets with an associated recovery user.
     *
     * @deprecated
     */
    createWalletsWithRecovery(body, options) {
        return this._client.post('/v1/wallets_with_recovery', { body, ...options });
    }
    /**
     * Get a wallet by wallet ID.
     *
     * @example
     * ```ts
     * const wallet = await client.wallets.get('wallet_id');
     * ```
     */
    get(walletID, options) {
        return this._client.get((0, path_1.path) `/v1/wallets/${walletID}`, options);
    }
}
exports.Wallets = Wallets;
Wallets.Transactions = transactions_1.Transactions;
Wallets.Balance = balance_1.Balance;
