// @privy-io/api-types/src/client.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

import type { RequestInit, RequestInfo, BodyInit } from './internal/builtin-types';
import type { HTTPMethod, PromiseOrValue, MergedRequestInit, FinalizedRequestInit } from './internal/types';
import { uuid4 } from './internal/utils/uuid';
import { validatePositiveInteger, isAbsoluteURL, safeJSON } from './internal/utils/values';
import { sleep } from './internal/utils/sleep';
export type { Logger, LogLevel } from './internal/utils/log';
import { castToError, isAbortError } from './internal/errors';
import type { APIResponseProps } from './internal/parse';
import { getPlatformHeaders } from './internal/detect-platform';
import * as Shims from './internal/shims';
import * as Opts from './internal/request-options';
import * as qs from './internal/qs';
import { VERSION } from './version';
import * as Errors from './core/error';
import * as Pagination from './core/pagination';
import { AbstractPage, type CursorParams, CursorResponse } from './core/pagination';
import * as Uploads from './core/uploads';
import * as API from './resources/index';
import { APIPromise } from './core/api-promise';
import {
  Aggregation,
  AggregationGroupBy,
  AggregationMethod,
  AggregationMetric,
  AggregationWindow,
  Aggregations,
} from './resources/aggregations';
import { Analytics, AnalyticsEventInput } from './resources/analytics';
import { AppResponse, Apps } from './resources/apps';
import {
  ClientAuth,
  CustomJwtAuthenticateRequestBody,
  CustomJwtLinkRequestBody,
  CustomOAuthProviderID,
  ExternalOAuthProviderID,
  FarcasterAuthenticateRequestBody,
  FarcasterConnectInitResponseBody,
  FarcasterConnectStatusCompletedResponseBody,
  FarcasterConnectStatusPendingResponseBody,
  FarcasterInitRequestBody,
  FarcasterLinkRequestBody,
  FarcasterSignerInitRequestBody,
  FarcasterSignerInitResponseBody,
  FarcasterSignerStatusResponseBody,
  FarcasterUnlinkRequestBody,
  FarcasterV2AuthenticateRequestBody,
  FarcasterV2InitRequestBody,
  FarcasterV2InitResponseBody,
  GuestAuthenticateRequestBody,
  MfaPasskeyEnrollmentRequestBody,
  MfaPasskeyInitRequestBody,
  MfaPasskeyInitResponseBody,
  MfaPasskeyVerifyRequestBody,
  MfaSMSEnrollRequestBody,
  MfaSMSInitRequestBody,
  MfaSMSVerifyRequestBody,
  MfaTotpInitResponseBody,
  MfaTotpInput,
  MfaVerifyResponseBody,
  OAuthAuthenticateRequestBody,
  OAuthAuthorizationCodeRequestBody,
  OAuthInitRequestBody,
  OAuthInitResponseBody,
  OAuthLinkRequestBody,
  OAuthLinkResponseBody,
  OAuthProviderID,
  OAuthUnlinkRequestBody,
  OAuthVerifyRequestBody,
  OAuthVerifyResponseBody,
  OptionalRefreshTokenInput,
  PasskeyAuthenticateRequestBody,
  PasskeyInitAuthenticateResponseBody,
  PasskeyInitLinkResponseBody,
  PasskeyInitRegisterResponseBody,
  PasskeyInitRequestBody,
  PasskeyLinkRequestBody,
  PasskeyRegisterRequestBody,
  PasskeyUnlinkRequestBody,
  PasswordlessAuthenticateRequestBody,
  PasswordlessInitRequestBody,
  PasswordlessLinkRequestBody,
  PasswordlessSMSAuthenticateRequestBody,
  PasswordlessSMSInitRequestBody,
  PasswordlessSMSLinkRequestBody,
  PasswordlessSMSUnlinkRequestBody,
  PasswordlessSMSUpdateRequestBody,
  PasswordlessUnlinkRequestBody,
  PasswordlessUpdateRequestBody,
  PrivyOAuthProviderID,
  SiweAuthenticateRequestBody,
  SiweInitRequestBody,
  SiweInitResponseBody,
  SiweLinkRequestBody,
  SiweLinkSmartWalletRequestBody,
  SiweUnlinkRequestBody,
  SiwsAuthenticateRequestBody,
  SiwsInitRequestBody,
  SiwsInitResponseBody,
  SiwsLinkRequestBody,
  SiwsUnlinkRequestBody,
  TelegramAuthenticateRequestBody,
  TelegramLinkRequestBody,
  TelegramUnlinkRequestBody,
} from './resources/client-auth';
import { CrossApp, CrossAppConnectionsResponse } from './resources/cross-app';
import {
  EmbeddedWallets,
  RecoveryConfigurationICloudRequestBody,
  RecoveryConfigurationICloudResponseBody,
  RecoveryKeyMaterialRequestBody,
  RecoveryKeyMaterialResponseBody,
  RecoveryOAuthAuthenticateResponseBody,
  RecoveryOAuthCallbackICloudExpoRequestBody,
  RecoveryOAuthInitICloudRequestBody,
  RecoveryOAuthInitRequestBody,
  SmartWalletConfiguration,
  SmartWalletConfigurationInput,
} from './resources/embedded-wallets';
import {
  CoinbaseOnRampInitInput,
  CoinbaseOnRampInitResponse,
  CoinbaseOnRampStatusResponse,
  Funding,
  MoonpayOnRampSignInput,
  MoonpayOnRampSignResponse,
} from './resources/funding';
import {
  KeyQuorum,
  KeyQuorumCreateParams,
  KeyQuorumDeleteParams,
  KeyQuorumDeleteResponse,
  KeyQuorumUpdateParams,
  KeyQuorums,
} from './resources/key-quorums';
import {
  Policies,
  Policy,
  PolicyCreateParams,
  PolicyCreateRuleParams,
  PolicyCreateRuleResponse,
  PolicyDeleteParams,
  PolicyDeleteResponse,
  PolicyDeleteRuleParams,
  PolicyDeleteRuleResponse,
  PolicyGetRuleParams,
  PolicyGetRuleResponse,
  PolicyUpdateParams,
  PolicyUpdateRuleParams,
  PolicyUpdateRuleResponse,
  SuiTransactionCommandCondition,
  SuiTransactionCommandOperator,
  SuiTransferObjectsCommandCondition,
  SuiTransferObjectsCommandField,
  TronTransactionCondition,
} from './resources/policies';
import {
  TransactionGetResponse,
  TransactionScanningRequestBody,
  TransactionScanningResponseBody,
  Transactions,
} from './resources/transactions';
import {
  AuthenticatedUser,
  CrossAppEmbeddedWallet,
  CrossAppSmartWallet,
  CustomMetadata,
  EmbeddedWalletRecoveryMethod,
  LinkedAccount,
  LinkedAccountAppleInput,
  LinkedAccountAppleOAuth,
  LinkedAccountAuthorizationKey,
  LinkedAccountBaseWallet,
  LinkedAccountBitcoinSegwitEmbeddedWallet,
  LinkedAccountBitcoinTaprootEmbeddedWallet,
  LinkedAccountCrossApp,
  LinkedAccountCurveSigningEmbeddedWallet,
  LinkedAccountCustomJwt,
  LinkedAccountCustomJwtInput,
  LinkedAccountCustomOAuth,
  LinkedAccountDiscordInput,
  LinkedAccountDiscordOAuth,
  LinkedAccountEmail,
  LinkedAccountEmailInput,
  LinkedAccountEmbeddedWallet,
  LinkedAccountEmbeddedWalletWithID,
  LinkedAccountEthereum,
  LinkedAccountEthereumEmbeddedWallet,
  LinkedAccountFarcaster,
  LinkedAccountFarcasterInput,
  LinkedAccountGitHubInput,
  LinkedAccountGitHubOAuth,
  LinkedAccountGoogleInput,
  LinkedAccountGoogleOAuth,
  LinkedAccountInput,
  LinkedAccountInstagramInput,
  LinkedAccountInstagramOAuth,
  LinkedAccountLineInput,
  LinkedAccountLineOAuth,
  LinkedAccountLinkedInInput,
  LinkedAccountLinkedInOAuth,
  LinkedAccountPasskey,
  LinkedAccountPhone,
  LinkedAccountPhoneInput,
  LinkedAccountSmartWallet,
  LinkedAccountSolana,
  LinkedAccountSolanaEmbeddedWallet,
  LinkedAccountSpotifyInput,
  LinkedAccountSpotifyOAuth,
  LinkedAccountTelegram,
  LinkedAccountTelegramInput,
  LinkedAccountTiktokInput,
  LinkedAccountTiktokOAuth,
  LinkedAccountTwitchInput,
  LinkedAccountTwitchOAuth,
  LinkedAccountTwitterInput,
  LinkedAccountTwitterOAuth,
  LinkedAccountType,
  LinkedAccountWalletInput,
  LinkedMfaMethod,
  OAuthTokens,
  PasskeyMfaMethod,
  SMSMfaMethod,
  SmartWalletType,
  TotpMfaMethod,
  User,
  UserBatchCreateInput,
  UserCreateParams,
  UserGetByCustomAuthIDParams,
  UserGetByDiscordUsernameParams,
  UserGetByEmailAddressParams,
  UserGetByFarcasterIDParams,
  UserGetByGitHubUsernameParams,
  UserGetByPhoneNumberParams,
  UserGetBySmartWalletAddressParams,
  UserGetByTelegramUserIDParams,
  UserGetByTelegramUsernameParams,
  UserGetByTwitterSubjectParams,
  UserGetByTwitterUsernameParams,
  UserGetByWalletAddressParams,
  UserListParams,
  UserPregenerateWalletsParams,
  UserSearchParams,
  UserSetCustomMetadataParams,
  UserUnlinkLinkedAccountParams,
  UserWithIdentityToken,
  Users,
  UsersCursor,
} from './resources/users';
import {
  CurveSigningChainType,
  CustodialWallet,
  CustodialWalletChainType,
  CustodialWalletCreateInput,
  CustodialWalletProvider,
  EthereumPersonalSignRpcInput,
  EthereumPersonalSignRpcResponse,
  EthereumSecp256k1SignRpcInput,
  EthereumSecp256k1SignRpcResponse,
  EthereumSendTransactionRpcInput,
  EthereumSendTransactionRpcResponse,
  EthereumSign7702AuthorizationRpcInput,
  EthereumSign7702AuthorizationRpcResponse,
  EthereumSignTransactionRpcInput,
  EthereumSignTransactionRpcResponse,
  EthereumSignTypedDataRpcInput,
  EthereumSignTypedDataRpcResponse,
  EthereumSignUserOperationRpcInput,
  EthereumSignUserOperationRpcResponse,
  ExtendedChainType,
  FirstClassChainType,
  HpkeImportConfig,
  SolanaSignAndSendTransactionRpcInput,
  SolanaSignAndSendTransactionRpcResponse,
  SolanaSignMessageRpcInput,
  SolanaSignMessageRpcResponse,
  SolanaSignTransactionRpcInput,
  SolanaSignTransactionRpcResponse,
  SuiCommandName,
  Wallet,
  WalletAuthenticateWithJwtParams,
  WalletAuthenticateWithJwtResponse,
  WalletChainType,
  WalletCreateParams,
  WalletCreateWalletsWithRecoveryParams,
  WalletCreateWalletsWithRecoveryResponse,
  WalletCustodian,
  WalletExportParams,
  WalletExportResponse,
  WalletInitImportParams,
  WalletInitImportResponse,
  WalletListParams,
  WalletRawSignParams,
  WalletRawSignResponse,
  WalletRevokeResponse,
  WalletRpcParams,
  WalletRpcResponse,
  WalletSubmitImportParams,
  WalletUpdateParams,
  Wallets,
  WalletsCursor,
} from './resources/wallets/wallets';
import { type Fetch } from './internal/builtin-types';
import { isRunningInBrowser } from './internal/detect-platform';
import { HeadersLike, NullableHeaders, buildHeaders } from './internal/headers';
import { FinalRequestOptions, RequestOptions } from './internal/request-options';
import { toBase64 } from './internal/utils/base64';
import { readEnv } from './internal/utils/env';
import {
  type LogLevel,
  type Logger,
  formatRequestDetails,
  loggerFor,
  parseLogLevel,
} from './internal/utils/log';
import { isEmptyObj } from './internal/utils/values';

const environments = {
  production: 'https://api.privy.io',
  staging: 'https://api.staging.privy.io',
};
type Environment = keyof typeof environments;

export interface ClientOptions {
  /**
   * App secret authentication.
   */
  appID?: string | undefined;

  /**
   * App secret authentication.
   */
  appSecret?: string | undefined;

  /**
   * Specifies the environment to use for the API.
   *
   * Each environment maps to a different base URL:
   * - `production` corresponds to `https://api.privy.io`
   * - `staging` corresponds to `https://api.staging.privy.io`
   */
  environment?: Environment | undefined;

  /**
   * Override the default base URL for the API, e.g., "https://api.example.com/v2/"
   *
   * Defaults to process.env['PRIVY_BASE_URL'].
   */
  baseURL?: string | null | undefined;

  /**
   * The maximum amount of time (in milliseconds) that the client should wait for a response
   * from the server before timing out a single request.
   *
   * Note that request timeouts are retried by default, so in a worst-case scenario you may wait
   * much longer than this timeout before the promise succeeds or fails.
   *
   * @unit milliseconds
   */
  timeout?: number | undefined;
  /**
   * Additional `RequestInit` options to be passed to `fetch` calls.
   * Properties will be overridden by per-request `fetchOptions`.
   */
  fetchOptions?: MergedRequestInit | undefined;

  /**
   * Specify a custom `fetch` function implementation.
   *
   * If not provided, we expect that `fetch` is defined globally.
   */
  fetch?: Fetch | undefined;

  /**
   * The maximum number of times that the client will retry a request in case of a
   * temporary failure, like a network error or a 5XX error from the server.
   *
   * @default 2
   */
  maxRetries?: number | undefined;

  /**
   * Default headers to include with every request to the API.
   *
   * These can be removed in individual requests by explicitly setting the
   * header to `null` in request options.
   */
  defaultHeaders?: HeadersLike | undefined;

  /**
   * Default query parameters to include with every request to the API.
   *
   * These can be removed in individual requests by explicitly setting the
   * param to `undefined` in request options.
   */
  defaultQuery?: Record<string, string | undefined> | undefined;

  /**
   * Set the log level.
   *
   * Defaults to process.env['PRIVY_LOG'] or 'warn' if it isn't set.
   */
  logLevel?: LogLevel | undefined;

  /**
   * Set the logger.
   *
   * Defaults to globalThis.console.
   */
  logger?: Logger | undefined;
}

/**
 * API Client for interfacing with the Privy API.
 */
export class Privy {
  appID: string;
  appSecret: string;

  baseURL: string;
  maxRetries: number;
  timeout: number;
  logger: Logger;
  logLevel: LogLevel | undefined;
  fetchOptions: MergedRequestInit | undefined;

  private fetch: Fetch;
  #encoder: Opts.RequestEncoder;
  protected idempotencyHeader?: string;
  private _options: ClientOptions;

  /**
   * API Client for interfacing with the Privy API.
   *
   * @param {string | undefined} [opts.appID=process.env['PRIVY_APP_ID'] ?? undefined]
   * @param {string | undefined} [opts.appSecret=process.env['PRIVY_APP_SECRET'] ?? undefined]
   * @param {Environment} [opts.environment=production] - Specifies the environment URL to use for the API.
   * @param {string} [opts.baseURL=process.env['PRIVY_BASE_URL'] ?? https://api.privy.io] - Override the default base URL for the API.
   * @param {number} [opts.timeout=1 minute] - The maximum amount of time (in milliseconds) the client will wait for a response before timing out.
   * @param {MergedRequestInit} [opts.fetchOptions] - Additional `RequestInit` options to be passed to `fetch` calls.
   * @param {Fetch} [opts.fetch] - Specify a custom `fetch` function implementation.
   * @param {number} [opts.maxRetries=2] - The maximum number of times the client will retry a request.
   * @param {HeadersLike} opts.defaultHeaders - Default headers to include with every request to the API.
   * @param {Record<string, string | undefined>} opts.defaultQuery - Default query parameters to include with every request to the API.
   */
  constructor({
    baseURL = readEnv('PRIVY_BASE_URL'),
    appID = readEnv('PRIVY_APP_ID'),
    appSecret = readEnv('PRIVY_APP_SECRET'),
    ...opts
  }: ClientOptions = {}) {
    if (appID === undefined) {
      throw new Errors.PrivyError(
        "The PRIVY_APP_ID environment variable is missing or empty; either provide it, or instantiate the Privy client with an appID option, like new Privy({ appID: 'My App ID' }).",
      );
    }
    if (appSecret === undefined) {
      throw new Errors.PrivyError(
        "The PRIVY_APP_SECRET environment variable is missing or empty; either provide it, or instantiate the Privy client with an appSecret option, like new Privy({ appSecret: 'My App Secret' }).",
      );
    }

    const options: ClientOptions = {
      appID,
      appSecret,
      ...opts,
      baseURL,
      environment: opts.environment ?? 'production',
    };

    if (isRunningInBrowser()) {
      throw new Errors.PrivyError(
        "It looks like you're running in a browser-like environment, which is disabled to protect your secret API credentials from attackers. If you have a strong business need for client-side use of this API, please open a GitHub issue with your use-case and security mitigations.",
      );
    }

    if (baseURL && opts.environment) {
      throw new Errors.PrivyError(
        'Ambiguous URL; The `baseURL` option (or PRIVY_BASE_URL env var) and the `environment` option are given. If you want to use the environment you must pass baseURL: null',
      );
    }

    this.baseURL = options.baseURL || environments[options.environment || 'production'];
    this.timeout = options.timeout ?? Privy.DEFAULT_TIMEOUT /* 1 minute */;
    this.logger = options.logger ?? console;
    const defaultLogLevel = 'warn';
    // Set default logLevel early so that we can log a warning in parseLogLevel.
    this.logLevel = defaultLogLevel;
    this.logLevel =
      parseLogLevel(options.logLevel, 'ClientOptions.logLevel', this) ??
      parseLogLevel(readEnv('PRIVY_LOG'), "process.env['PRIVY_LOG']", this) ??
      defaultLogLevel;
    this.fetchOptions = options.fetchOptions;
    this.maxRetries = options.maxRetries ?? 2;
    this.fetch = options.fetch ?? Shims.getDefaultFetch();
    this.#encoder = Opts.FallbackEncoder;

    this._options = options;

    this.appID = appID;
    this.appSecret = appSecret;
  }

  /**
   * Create a new client instance re-using the same options given to the current client with optional overriding.
   */
  withOptions(options: Partial<ClientOptions>): this {
    const client = new (this.constructor as any as new (props: ClientOptions) => typeof this)({
      ...this._options,
      environment: options.environment ? options.environment : undefined,
      baseURL: options.environment ? undefined : this.baseURL,
      maxRetries: this.maxRetries,
      timeout: this.timeout,
      logger: this.logger,
      logLevel: this.logLevel,
      fetch: this.fetch,
      fetchOptions: this.fetchOptions,
      appID: this.appID,
      appSecret: this.appSecret,
      ...options,
    });
    return client;
  }

  /**
   * Check whether the base URL is set to its default.
   */
  #baseURLOverridden(): boolean {
    return this.baseURL !== environments[this._options.environment || 'production'];
  }

  protected defaultQuery(): Record<string, string | undefined> | undefined {
    return this._options.defaultQuery;
  }

  protected validateHeaders({ values, nulls }: NullableHeaders) {
    return;
  }

  protected async authHeaders(opts: FinalRequestOptions): Promise<NullableHeaders | undefined> {
    if (!this.appID) {
      return undefined;
    }

    if (!this.appSecret) {
      return undefined;
    }

    const credentials = `${this.appID}:${this.appSecret}`;
    const Authorization = `Basic ${toBase64(credentials)}`;
    return buildHeaders([{ Authorization }]);
  }

  protected stringifyQuery(query: Record<string, unknown>): string {
    return qs.stringify(query, { arrayFormat: 'comma' });
  }

  private getUserAgent(): string {
    return `${this.constructor.name}/JS ${VERSION}`;
  }

  protected defaultIdempotencyKey(): string {
    return `stainless-node-retry-${uuid4()}`;
  }

  protected makeStatusError(
    status: number,
    error: Object,
    message: string | undefined,
    headers: Headers,
  ): Errors.APIError {
    return Errors.APIError.generate(status, error, message, headers);
  }

  buildURL(
    path: string,
    query: Record<string, unknown> | null | undefined,
    defaultBaseURL?: string | undefined,
  ): string {
    const baseURL = (!this.#baseURLOverridden() && defaultBaseURL) || this.baseURL;
    const url =
      isAbsoluteURL(path) ?
        new URL(path)
      : new URL(baseURL + (baseURL.endsWith('/') && path.startsWith('/') ? path.slice(1) : path));

    const defaultQuery = this.defaultQuery();
    if (!isEmptyObj(defaultQuery)) {
      query = { ...defaultQuery, ...query };
    }

    if (typeof query === 'object' && query && !Array.isArray(query)) {
      url.search = this.stringifyQuery(query as Record<string, unknown>);
    }

    return url.toString();
  }

  /**
   * Used as a callback for mutating the given `FinalRequestOptions` object.
   */
  protected async prepareOptions(options: FinalRequestOptions): Promise<void> {}

  /**
   * Used as a callback for mutating the given `RequestInit` object.
   *
   * This is useful for cases where you want to add certain headers based off of
   * the request properties, e.g. `method` or `url`.
   */
  protected async prepareRequest(
    request: RequestInit,
    { url, options }: { url: string; options: FinalRequestOptions },
  ): Promise<void> {}

  get<Rsp>(path: string, opts?: PromiseOrValue<RequestOptions>): APIPromise<Rsp> {
    return this.methodRequest('get', path, opts);
  }

  post<Rsp>(path: string, opts?: PromiseOrValue<RequestOptions>): APIPromise<Rsp> {
    return this.methodRequest('post', path, opts);
  }

  patch<Rsp>(path: string, opts?: PromiseOrValue<RequestOptions>): APIPromise<Rsp> {
    return this.methodRequest('patch', path, opts);
  }

  put<Rsp>(path: string, opts?: PromiseOrValue<RequestOptions>): APIPromise<Rsp> {
    return this.methodRequest('put', path, opts);
  }

  delete<Rsp>(path: string, opts?: PromiseOrValue<RequestOptions>): APIPromise<Rsp> {
    return this.methodRequest('delete', path, opts);
  }

  private methodRequest<Rsp>(
    method: HTTPMethod,
    path: string,
    opts?: PromiseOrValue<RequestOptions>,
  ): APIPromise<Rsp> {
    return this.request(
      Promise.resolve(opts).then((opts) => {
        return { method, path, ...opts };
      }),
    );
  }

  request<Rsp>(
    options: PromiseOrValue<FinalRequestOptions>,
    remainingRetries: number | null = null,
  ): APIPromise<Rsp> {
    return new APIPromise(this, this.makeRequest(options, remainingRetries, undefined));
  }

  private async makeRequest(
    optionsInput: PromiseOrValue<FinalRequestOptions>,
    retriesRemaining: number | null,
    retryOfRequestLogID: string | undefined,
  ): Promise<APIResponseProps> {
    const options = await optionsInput;
    const maxRetries = options.maxRetries ?? this.maxRetries;
    if (retriesRemaining == null) {
      retriesRemaining = maxRetries;
    }

    await this.prepareOptions(options);

    const { req, url, timeout } = await this.buildRequest(options, {
      retryCount: maxRetries - retriesRemaining,
    });

    await this.prepareRequest(req, { url, options });

    /** Not an API request ID, just for correlating local log entries. */
    const requestLogID = 'log_' + ((Math.random() * (1 << 24)) | 0).toString(16).padStart(6, '0');
    const retryLogStr = retryOfRequestLogID === undefined ? '' : `, retryOf: ${retryOfRequestLogID}`;
    const startTime = Date.now();

    loggerFor(this).debug(
      `[${requestLogID}] sending request`,
      formatRequestDetails({
        retryOfRequestLogID,
        method: options.method,
        url,
        options,
        headers: req.headers,
      }),
    );

    if (options.signal?.aborted) {
      throw new Errors.APIUserAbortError();
    }

    const controller = new AbortController();
    const response = await this.fetchWithTimeout(url, req, timeout, controller).catch(castToError);
    const headersTime = Date.now();

    if (response instanceof globalThis.Error) {
      const retryMessage = `retrying, ${retriesRemaining} attempts remaining`;
      if (options.signal?.aborted) {
        throw new Errors.APIUserAbortError();
      }
      // detect native connection timeout errors
      // deno throws "TypeError: error sending request for url (https://example/): client error (Connect): tcp connect error: Operation timed out (os error 60): Operation timed out (os error 60)"
      // undici throws "TypeError: fetch failed" with cause "ConnectTimeoutError: Connect Timeout Error (attempted address: example:443, timeout: 1ms)"
      // others do not provide enough information to distinguish timeouts from other connection errors
      const isTimeout =
        isAbortError(response) ||
        /timed? ?out/i.test(String(response) + ('cause' in response ? String(response.cause) : ''));
      if (retriesRemaining) {
        loggerFor(this).info(
          `[${requestLogID}] connection ${isTimeout ? 'timed out' : 'failed'} - ${retryMessage}`,
        );
        loggerFor(this).debug(
          `[${requestLogID}] connection ${isTimeout ? 'timed out' : 'failed'} (${retryMessage})`,
          formatRequestDetails({
            retryOfRequestLogID,
            url,
            durationMs: headersTime - startTime,
            message: response.message,
          }),
        );
        return this.retryRequest(options, retriesRemaining, retryOfRequestLogID ?? requestLogID);
      }
      loggerFor(this).info(
        `[${requestLogID}] connection ${isTimeout ? 'timed out' : 'failed'} - error; no more retries left`,
      );
      loggerFor(this).debug(
        `[${requestLogID}] connection ${isTimeout ? 'timed out' : 'failed'} (error; no more retries left)`,
        formatRequestDetails({
          retryOfRequestLogID,
          url,
          durationMs: headersTime - startTime,
          message: response.message,
        }),
      );
      if (isTimeout) {
        throw new Errors.APIConnectionTimeoutError();
      }
      throw new Errors.APIConnectionError({ cause: response });
    }

    const responseInfo = `[${requestLogID}${retryLogStr}] ${req.method} ${url} ${
      response.ok ? 'succeeded' : 'failed'
    } with status ${response.status} in ${headersTime - startTime}ms`;

    if (!response.ok) {
      const shouldRetry = await this.shouldRetry(response);
      if (retriesRemaining && shouldRetry) {
        const retryMessage = `retrying, ${retriesRemaining} attempts remaining`;

        // We don't need the body of this response.
        await Shims.CancelReadableStream(response.body);
        loggerFor(this).info(`${responseInfo} - ${retryMessage}`);
        loggerFor(this).debug(
          `[${requestLogID}] response error (${retryMessage})`,
          formatRequestDetails({
            retryOfRequestLogID,
            url: response.url,
            status: response.status,
            headers: response.headers,
            durationMs: headersTime - startTime,
          }),
        );
        return this.retryRequest(
          options,
          retriesRemaining,
          retryOfRequestLogID ?? requestLogID,
          response.headers,
        );
      }

      const retryMessage = shouldRetry ? `error; no more retries left` : `error; not retryable`;

      loggerFor(this).info(`${responseInfo} - ${retryMessage}`);

      const errText = await response.text().catch((err: any) => castToError(err).message);
      const errJSON = safeJSON(errText);
      const errMessage = errJSON ? undefined : errText;

      loggerFor(this).debug(
        `[${requestLogID}] response error (${retryMessage})`,
        formatRequestDetails({
          retryOfRequestLogID,
          url: response.url,
          status: response.status,
          headers: response.headers,
          message: errMessage,
          durationMs: Date.now() - startTime,
        }),
      );

      const err = this.makeStatusError(response.status, errJSON, errMessage, response.headers);
      throw err;
    }

    loggerFor(this).info(responseInfo);
    loggerFor(this).debug(
      `[${requestLogID}] response start`,
      formatRequestDetails({
        retryOfRequestLogID,
        url: response.url,
        status: response.status,
        headers: response.headers,
        durationMs: headersTime - startTime,
      }),
    );

    return { response, options, controller, requestLogID, retryOfRequestLogID, startTime };
  }

  getAPIList<Item, PageClass extends Pagination.AbstractPage<Item> = Pagination.AbstractPage<Item>>(
    path: string,
    Page: new (...args: any[]) => PageClass,
    opts?: RequestOptions,
  ): Pagination.PagePromise<PageClass, Item> {
    return this.requestAPIList(Page, { method: 'get', path, ...opts });
  }

  requestAPIList<
    Item = unknown,
    PageClass extends Pagination.AbstractPage<Item> = Pagination.AbstractPage<Item>,
  >(
    Page: new (...args: ConstructorParameters<typeof Pagination.AbstractPage>) => PageClass,
    options: FinalRequestOptions,
  ): Pagination.PagePromise<PageClass, Item> {
    const request = this.makeRequest(options, null, undefined);
    return new Pagination.PagePromise<PageClass, Item>(this as any as Privy, request, Page);
  }

  async fetchWithTimeout(
    url: RequestInfo,
    init: RequestInit | undefined,
    ms: number,
    controller: AbortController,
  ): Promise<Response> {
    const { signal, method, ...options } = init || {};
    if (signal) signal.addEventListener('abort', () => controller.abort());

    const timeout = setTimeout(() => controller.abort(), ms);

    const isReadableBody =
      ((globalThis as any).ReadableStream && options.body instanceof (globalThis as any).ReadableStream) ||
      (typeof options.body === 'object' && options.body !== null && Symbol.asyncIterator in options.body);

    const fetchOptions: RequestInit = {
      signal: controller.signal as any,
      ...(isReadableBody ? { duplex: 'half' } : {}),
      method: 'GET',
      ...options,
    };
    if (method) {
      // Custom methods like 'patch' need to be uppercased
      // See https://github.com/nodejs/undici/issues/2294
      fetchOptions.method = method.toUpperCase();
    }

    try {
      // use undefined this binding; fetch errors if bound to something else in browser/cloudflare
      return await this.fetch.call(undefined, url, fetchOptions);
    } finally {
      clearTimeout(timeout);
    }
  }

  private async shouldRetry(response: Response): Promise<boolean> {
    // Note this is not a standard header.
    const shouldRetryHeader = response.headers.get('x-should-retry');

    // If the server explicitly says whether or not to retry, obey.
    if (shouldRetryHeader === 'true') return true;
    if (shouldRetryHeader === 'false') return false;

    // Retry on request timeouts.
    if (response.status === 408) return true;

    // Retry on lock timeouts.
    if (response.status === 409) return true;

    // Retry on rate limits.
    if (response.status === 429) return true;

    // Retry internal errors.
    if (response.status >= 500) return true;

    return false;
  }

  private async retryRequest(
    options: FinalRequestOptions,
    retriesRemaining: number,
    requestLogID: string,
    responseHeaders?: Headers | undefined,
  ): Promise<APIResponseProps> {
    let timeoutMillis: number | undefined;

    // Note the `retry-after-ms` header may not be standard, but is a good idea and we'd like proactive support for it.
    const retryAfterMillisHeader = responseHeaders?.get('retry-after-ms');
    if (retryAfterMillisHeader) {
      const timeoutMs = parseFloat(retryAfterMillisHeader);
      if (!Number.isNaN(timeoutMs)) {
        timeoutMillis = timeoutMs;
      }
    }

    // About the Retry-After header: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After
    const retryAfterHeader = responseHeaders?.get('retry-after');
    if (retryAfterHeader && !timeoutMillis) {
      const timeoutSeconds = parseFloat(retryAfterHeader);
      if (!Number.isNaN(timeoutSeconds)) {
        timeoutMillis = timeoutSeconds * 1000;
      } else {
        timeoutMillis = Date.parse(retryAfterHeader) - Date.now();
      }
    }

    // If the API asks us to wait a certain amount of time (and it's a reasonable amount),
    // just do what it says, but otherwise calculate a default
    if (!(timeoutMillis && 0 <= timeoutMillis && timeoutMillis < 60 * 1000)) {
      const maxRetries = options.maxRetries ?? this.maxRetries;
      timeoutMillis = this.calculateDefaultRetryTimeoutMillis(retriesRemaining, maxRetries);
    }
    await sleep(timeoutMillis);

    return this.makeRequest(options, retriesRemaining - 1, requestLogID);
  }

  private calculateDefaultRetryTimeoutMillis(retriesRemaining: number, maxRetries: number): number {
    const initialRetryDelay = 0.5;
    const maxRetryDelay = 8.0;

    const numRetries = maxRetries - retriesRemaining;

    // Apply exponential backoff, but not more than the max.
    const sleepSeconds = Math.min(initialRetryDelay * Math.pow(2, numRetries), maxRetryDelay);

    // Apply some jitter, take up to at most 25 percent of the retry time.
    const jitter = 1 - Math.random() * 0.25;

    return sleepSeconds * jitter * 1000;
  }

  async buildRequest(
    inputOptions: FinalRequestOptions,
    { retryCount = 0 }: { retryCount?: number } = {},
  ): Promise<{ req: FinalizedRequestInit; url: string; timeout: number }> {
    const options = { ...inputOptions };
    const { method, path, query, defaultBaseURL } = options;

    const url = this.buildURL(path!, query as Record<string, unknown>, defaultBaseURL);
    if ('timeout' in options) validatePositiveInteger('timeout', options.timeout);
    options.timeout = options.timeout ?? this.timeout;
    const { bodyHeaders, body } = this.buildBody({ options });
    const reqHeaders = await this.buildHeaders({ options: inputOptions, method, bodyHeaders, retryCount });

    const req: FinalizedRequestInit = {
      method,
      headers: reqHeaders,
      ...(options.signal && { signal: options.signal }),
      ...((globalThis as any).ReadableStream &&
        body instanceof (globalThis as any).ReadableStream && { duplex: 'half' }),
      ...(body && { body }),
      ...((this.fetchOptions as any) ?? {}),
      ...((options.fetchOptions as any) ?? {}),
    };

    return { req, url, timeout: options.timeout };
  }

  private async buildHeaders({
    options,
    method,
    bodyHeaders,
    retryCount,
  }: {
    options: FinalRequestOptions;
    method: HTTPMethod;
    bodyHeaders: HeadersLike;
    retryCount: number;
  }): Promise<Headers> {
    let idempotencyHeaders: HeadersLike = {};
    if (this.idempotencyHeader && method !== 'get') {
      if (!options.idempotencyKey) options.idempotencyKey = this.defaultIdempotencyKey();
      idempotencyHeaders[this.idempotencyHeader] = options.idempotencyKey;
    }

    const headers = buildHeaders([
      idempotencyHeaders,
      {
        Accept: 'application/json',
        'User-Agent': this.getUserAgent(),
        'X-Stainless-Retry-Count': String(retryCount),
        ...(options.timeout ? { 'X-Stainless-Timeout': String(Math.trunc(options.timeout / 1000)) } : {}),
        ...getPlatformHeaders(),
        'privy-app-id': this.appID,
      },
      await this.authHeaders(options),
      this._options.defaultHeaders,
      bodyHeaders,
      options.headers,
    ]);

    this.validateHeaders(headers);

    return headers.values;
  }

  private buildBody({ options: { body, headers: rawHeaders } }: { options: FinalRequestOptions }): {
    bodyHeaders: HeadersLike;
    body: BodyInit | undefined;
  } {
    if (!body) {
      return { bodyHeaders: undefined, body: undefined };
    }
    const headers = buildHeaders([rawHeaders]);
    if (
      // Pass raw type verbatim
      ArrayBuffer.isView(body) ||
      body instanceof ArrayBuffer ||
      body instanceof DataView ||
      (typeof body === 'string' &&
        // Preserve legacy string encoding behavior for now
        headers.values.has('content-type')) ||
      // `Blob` is superset of `File`
      ((globalThis as any).Blob && body instanceof (globalThis as any).Blob) ||
      // `FormData` -> `multipart/form-data`
      body instanceof FormData ||
      // `URLSearchParams` -> `application/x-www-form-urlencoded`
      body instanceof URLSearchParams ||
      // Send chunked stream (each chunk has own `length`)
      ((globalThis as any).ReadableStream && body instanceof (globalThis as any).ReadableStream)
    ) {
      return { bodyHeaders: undefined, body: body as BodyInit };
    } else if (
      typeof body === 'object' &&
      (Symbol.asyncIterator in body ||
        (Symbol.iterator in body && 'next' in body && typeof body.next === 'function'))
    ) {
      return { bodyHeaders: undefined, body: Shims.ReadableStreamFrom(body as AsyncIterable<Uint8Array>) };
    } else {
      return this.#encoder({ body, headers });
    }
  }

  static Privy = this;
  static DEFAULT_TIMEOUT = 60000; // 1 minute

  static PrivyError = Errors.PrivyError;
  static APIError = Errors.APIError;
  static APIConnectionError = Errors.APIConnectionError;
  static APIConnectionTimeoutError = Errors.APIConnectionTimeoutError;
  static APIUserAbortError = Errors.APIUserAbortError;
  static NotFoundError = Errors.NotFoundError;
  static ConflictError = Errors.ConflictError;
  static RateLimitError = Errors.RateLimitError;
  static BadRequestError = Errors.BadRequestError;
  static AuthenticationError = Errors.AuthenticationError;
  static InternalServerError = Errors.InternalServerError;
  static PermissionDeniedError = Errors.PermissionDeniedError;
  static UnprocessableEntityError = Errors.UnprocessableEntityError;

  static toFile = Uploads.toFile;

  wallets: API.Wallets = new API.Wallets(this);
  apps: API.Apps = new API.Apps(this);
  embeddedWallets: API.EmbeddedWallets = new API.EmbeddedWallets(this);
  clientAuth: API.ClientAuth = new API.ClientAuth(this);
  crossApp: API.CrossApp = new API.CrossApp(this);
  funding: API.Funding = new API.Funding(this);
  users: API.Users = new API.Users(this);
  analytics: API.Analytics = new API.Analytics(this);
  policies: API.Policies = new API.Policies(this);
  transactions: API.Transactions = new API.Transactions(this);
  keyQuorums: API.KeyQuorums = new API.KeyQuorums(this);
  aggregations: API.Aggregations = new API.Aggregations(this);
}

Privy.Wallets = Wallets;
Privy.Apps = Apps;
Privy.EmbeddedWallets = EmbeddedWallets;
Privy.ClientAuth = ClientAuth;
Privy.CrossApp = CrossApp;
Privy.Funding = Funding;
Privy.Users = Users;
Privy.Analytics = Analytics;
Privy.Policies = Policies;
Privy.Transactions = Transactions;
Privy.KeyQuorums = KeyQuorums;
Privy.Aggregations = Aggregations;

export declare namespace Privy {
  export type RequestOptions = Opts.RequestOptions;

  export import Cursor = Pagination.Cursor;
  export { type CursorParams as CursorParams, type CursorResponse as CursorResponse };

  export {
    Wallets as Wallets,
    type CurveSigningChainType as CurveSigningChainType,
    type FirstClassChainType as FirstClassChainType,
    type Wallet as Wallet,
    type WalletChainType as WalletChainType,
    type WalletRevokeResponse as WalletRevokeResponse,
    type ExtendedChainType as ExtendedChainType,
    type WalletCustodian as WalletCustodian,
    type CustodialWalletProvider as CustodialWalletProvider,
    type CustodialWalletChainType as CustodialWalletChainType,
    type CustodialWalletCreateInput as CustodialWalletCreateInput,
    type CustodialWallet as CustodialWallet,
    type HpkeImportConfig as HpkeImportConfig,
    type SuiCommandName as SuiCommandName,
    type EthereumPersonalSignRpcInput as EthereumPersonalSignRpcInput,
    type EthereumSignTransactionRpcInput as EthereumSignTransactionRpcInput,
    type EthereumSendTransactionRpcInput as EthereumSendTransactionRpcInput,
    type EthereumSignTypedDataRpcInput as EthereumSignTypedDataRpcInput,
    type EthereumSignUserOperationRpcInput as EthereumSignUserOperationRpcInput,
    type EthereumSign7702AuthorizationRpcInput as EthereumSign7702AuthorizationRpcInput,
    type EthereumSecp256k1SignRpcInput as EthereumSecp256k1SignRpcInput,
    type SolanaSignTransactionRpcInput as SolanaSignTransactionRpcInput,
    type SolanaSignAndSendTransactionRpcInput as SolanaSignAndSendTransactionRpcInput,
    type SolanaSignMessageRpcInput as SolanaSignMessageRpcInput,
    type EthereumSignTransactionRpcResponse as EthereumSignTransactionRpcResponse,
    type EthereumSendTransactionRpcResponse as EthereumSendTransactionRpcResponse,
    type EthereumPersonalSignRpcResponse as EthereumPersonalSignRpcResponse,
    type EthereumSignTypedDataRpcResponse as EthereumSignTypedDataRpcResponse,
    type EthereumSignUserOperationRpcResponse as EthereumSignUserOperationRpcResponse,
    type EthereumSign7702AuthorizationRpcResponse as EthereumSign7702AuthorizationRpcResponse,
    type EthereumSecp256k1SignRpcResponse as EthereumSecp256k1SignRpcResponse,
    type SolanaSignTransactionRpcResponse as SolanaSignTransactionRpcResponse,
    type SolanaSignAndSendTransactionRpcResponse as SolanaSignAndSendTransactionRpcResponse,
    type SolanaSignMessageRpcResponse as SolanaSignMessageRpcResponse,
    type WalletExportResponse as WalletExportResponse,
    type WalletInitImportResponse as WalletInitImportResponse,
    type WalletRawSignResponse as WalletRawSignResponse,
    type WalletRpcResponse as WalletRpcResponse,
    type WalletAuthenticateWithJwtResponse as WalletAuthenticateWithJwtResponse,
    type WalletCreateWalletsWithRecoveryResponse as WalletCreateWalletsWithRecoveryResponse,
    type WalletsCursor as WalletsCursor,
    type WalletCreateParams as WalletCreateParams,
    type WalletListParams as WalletListParams,
    type WalletExportParams as WalletExportParams,
    type WalletInitImportParams as WalletInitImportParams,
    type WalletRawSignParams as WalletRawSignParams,
    type WalletRpcParams as WalletRpcParams,
    type WalletSubmitImportParams as WalletSubmitImportParams,
    type WalletUpdateParams as WalletUpdateParams,
    type WalletAuthenticateWithJwtParams as WalletAuthenticateWithJwtParams,
    type WalletCreateWalletsWithRecoveryParams as WalletCreateWalletsWithRecoveryParams,
  };

  export { Apps as Apps, type AppResponse as AppResponse };

  export {
    EmbeddedWallets as EmbeddedWallets,
    type RecoveryConfigurationICloudRequestBody as RecoveryConfigurationICloudRequestBody,
    type RecoveryConfigurationICloudResponseBody as RecoveryConfigurationICloudResponseBody,
    type RecoveryKeyMaterialRequestBody as RecoveryKeyMaterialRequestBody,
    type RecoveryKeyMaterialResponseBody as RecoveryKeyMaterialResponseBody,
    type RecoveryOAuthAuthenticateResponseBody as RecoveryOAuthAuthenticateResponseBody,
    type RecoveryOAuthCallbackICloudExpoRequestBody as RecoveryOAuthCallbackICloudExpoRequestBody,
    type RecoveryOAuthInitICloudRequestBody as RecoveryOAuthInitICloudRequestBody,
    type RecoveryOAuthInitRequestBody as RecoveryOAuthInitRequestBody,
    type SmartWalletConfiguration as SmartWalletConfiguration,
    type SmartWalletConfigurationInput as SmartWalletConfigurationInput,
  };

  export {
    ClientAuth as ClientAuth,
    type CustomJwtAuthenticateRequestBody as CustomJwtAuthenticateRequestBody,
    type CustomJwtLinkRequestBody as CustomJwtLinkRequestBody,
    type FarcasterAuthenticateRequestBody as FarcasterAuthenticateRequestBody,
    type FarcasterConnectInitResponseBody as FarcasterConnectInitResponseBody,
    type FarcasterConnectStatusCompletedResponseBody as FarcasterConnectStatusCompletedResponseBody,
    type FarcasterConnectStatusPendingResponseBody as FarcasterConnectStatusPendingResponseBody,
    type FarcasterInitRequestBody as FarcasterInitRequestBody,
    type FarcasterLinkRequestBody as FarcasterLinkRequestBody,
    type FarcasterSignerInitRequestBody as FarcasterSignerInitRequestBody,
    type FarcasterSignerInitResponseBody as FarcasterSignerInitResponseBody,
    type FarcasterSignerStatusResponseBody as FarcasterSignerStatusResponseBody,
    type FarcasterUnlinkRequestBody as FarcasterUnlinkRequestBody,
    type FarcasterV2AuthenticateRequestBody as FarcasterV2AuthenticateRequestBody,
    type FarcasterV2InitRequestBody as FarcasterV2InitRequestBody,
    type FarcasterV2InitResponseBody as FarcasterV2InitResponseBody,
    type GuestAuthenticateRequestBody as GuestAuthenticateRequestBody,
    type MfaPasskeyEnrollmentRequestBody as MfaPasskeyEnrollmentRequestBody,
    type MfaPasskeyInitRequestBody as MfaPasskeyInitRequestBody,
    type MfaPasskeyInitResponseBody as MfaPasskeyInitResponseBody,
    type MfaPasskeyVerifyRequestBody as MfaPasskeyVerifyRequestBody,
    type MfaSMSEnrollRequestBody as MfaSMSEnrollRequestBody,
    type MfaSMSInitRequestBody as MfaSMSInitRequestBody,
    type MfaSMSVerifyRequestBody as MfaSMSVerifyRequestBody,
    type MfaTotpInitResponseBody as MfaTotpInitResponseBody,
    type MfaTotpInput as MfaTotpInput,
    type MfaVerifyResponseBody as MfaVerifyResponseBody,
    type OAuthAuthenticateRequestBody as OAuthAuthenticateRequestBody,
    type OAuthAuthorizationCodeRequestBody as OAuthAuthorizationCodeRequestBody,
    type OAuthInitRequestBody as OAuthInitRequestBody,
    type OAuthInitResponseBody as OAuthInitResponseBody,
    type OAuthLinkRequestBody as OAuthLinkRequestBody,
    type OAuthLinkResponseBody as OAuthLinkResponseBody,
    type OAuthUnlinkRequestBody as OAuthUnlinkRequestBody,
    type OAuthVerifyRequestBody as OAuthVerifyRequestBody,
    type OAuthVerifyResponseBody as OAuthVerifyResponseBody,
    type OptionalRefreshTokenInput as OptionalRefreshTokenInput,
    type PasskeyAuthenticateRequestBody as PasskeyAuthenticateRequestBody,
    type PasskeyInitAuthenticateResponseBody as PasskeyInitAuthenticateResponseBody,
    type PasskeyInitLinkResponseBody as PasskeyInitLinkResponseBody,
    type PasskeyInitRegisterResponseBody as PasskeyInitRegisterResponseBody,
    type PasskeyInitRequestBody as PasskeyInitRequestBody,
    type PasskeyLinkRequestBody as PasskeyLinkRequestBody,
    type PasskeyRegisterRequestBody as PasskeyRegisterRequestBody,
    type PasskeyUnlinkRequestBody as PasskeyUnlinkRequestBody,
    type PasswordlessAuthenticateRequestBody as PasswordlessAuthenticateRequestBody,
    type PasswordlessInitRequestBody as PasswordlessInitRequestBody,
    type PasswordlessLinkRequestBody as PasswordlessLinkRequestBody,
    type PasswordlessSMSAuthenticateRequestBody as PasswordlessSMSAuthenticateRequestBody,
    type PasswordlessSMSInitRequestBody as PasswordlessSMSInitRequestBody,
    type PasswordlessSMSLinkRequestBody as PasswordlessSMSLinkRequestBody,
    type PasswordlessSMSUnlinkRequestBody as PasswordlessSMSUnlinkRequestBody,
    type PasswordlessSMSUpdateRequestBody as PasswordlessSMSUpdateRequestBody,
    type PasswordlessUnlinkRequestBody as PasswordlessUnlinkRequestBody,
    type PasswordlessUpdateRequestBody as PasswordlessUpdateRequestBody,
    type SiweAuthenticateRequestBody as SiweAuthenticateRequestBody,
    type SiweInitRequestBody as SiweInitRequestBody,
    type SiweInitResponseBody as SiweInitResponseBody,
    type SiweLinkRequestBody as SiweLinkRequestBody,
    type SiweLinkSmartWalletRequestBody as SiweLinkSmartWalletRequestBody,
    type SiweUnlinkRequestBody as SiweUnlinkRequestBody,
    type SiwsAuthenticateRequestBody as SiwsAuthenticateRequestBody,
    type SiwsInitRequestBody as SiwsInitRequestBody,
    type SiwsInitResponseBody as SiwsInitResponseBody,
    type SiwsLinkRequestBody as SiwsLinkRequestBody,
    type SiwsUnlinkRequestBody as SiwsUnlinkRequestBody,
    type TelegramAuthenticateRequestBody as TelegramAuthenticateRequestBody,
    type TelegramLinkRequestBody as TelegramLinkRequestBody,
    type TelegramUnlinkRequestBody as TelegramUnlinkRequestBody,
    type ExternalOAuthProviderID as ExternalOAuthProviderID,
    type PrivyOAuthProviderID as PrivyOAuthProviderID,
    type CustomOAuthProviderID as CustomOAuthProviderID,
    type OAuthProviderID as OAuthProviderID,
  };

  export { CrossApp as CrossApp, type CrossAppConnectionsResponse as CrossAppConnectionsResponse };

  export {
    Funding as Funding,
    type CoinbaseOnRampInitInput as CoinbaseOnRampInitInput,
    type CoinbaseOnRampInitResponse as CoinbaseOnRampInitResponse,
    type CoinbaseOnRampStatusResponse as CoinbaseOnRampStatusResponse,
    type MoonpayOnRampSignInput as MoonpayOnRampSignInput,
    type MoonpayOnRampSignResponse as MoonpayOnRampSignResponse,
  };

  export {
    Users as Users,
    type AuthenticatedUser as AuthenticatedUser,
    type LinkedAccount as LinkedAccount,
    type User as User,
    type UserWithIdentityToken as UserWithIdentityToken,
    type LinkedAccountEmail as LinkedAccountEmail,
    type LinkedAccountPhone as LinkedAccountPhone,
    type LinkedAccountBaseWallet as LinkedAccountBaseWallet,
    type LinkedAccountEthereum as LinkedAccountEthereum,
    type SmartWalletType as SmartWalletType,
    type LinkedAccountSmartWallet as LinkedAccountSmartWallet,
    type LinkedAccountSolana as LinkedAccountSolana,
    type LinkedAccountFarcaster as LinkedAccountFarcaster,
    type LinkedAccountPasskey as LinkedAccountPasskey,
    type LinkedAccountTelegram as LinkedAccountTelegram,
    type EmbeddedWalletRecoveryMethod as EmbeddedWalletRecoveryMethod,
    type LinkedAccountEthereumEmbeddedWallet as LinkedAccountEthereumEmbeddedWallet,
    type LinkedAccountSolanaEmbeddedWallet as LinkedAccountSolanaEmbeddedWallet,
    type LinkedAccountBitcoinSegwitEmbeddedWallet as LinkedAccountBitcoinSegwitEmbeddedWallet,
    type LinkedAccountBitcoinTaprootEmbeddedWallet as LinkedAccountBitcoinTaprootEmbeddedWallet,
    type LinkedAccountCurveSigningEmbeddedWallet as LinkedAccountCurveSigningEmbeddedWallet,
    type LinkedAccountEmbeddedWallet as LinkedAccountEmbeddedWallet,
    type LinkedAccountEmbeddedWalletWithID as LinkedAccountEmbeddedWalletWithID,
    type LinkedAccountGoogleOAuth as LinkedAccountGoogleOAuth,
    type LinkedAccountTwitterOAuth as LinkedAccountTwitterOAuth,
    type LinkedAccountDiscordOAuth as LinkedAccountDiscordOAuth,
    type LinkedAccountGitHubOAuth as LinkedAccountGitHubOAuth,
    type LinkedAccountLinkedInOAuth as LinkedAccountLinkedInOAuth,
    type LinkedAccountSpotifyOAuth as LinkedAccountSpotifyOAuth,
    type LinkedAccountInstagramOAuth as LinkedAccountInstagramOAuth,
    type LinkedAccountTiktokOAuth as LinkedAccountTiktokOAuth,
    type LinkedAccountLineOAuth as LinkedAccountLineOAuth,
    type LinkedAccountTwitchOAuth as LinkedAccountTwitchOAuth,
    type LinkedAccountAppleOAuth as LinkedAccountAppleOAuth,
    type LinkedAccountCustomOAuth as LinkedAccountCustomOAuth,
    type LinkedAccountCustomJwt as LinkedAccountCustomJwt,
    type CrossAppEmbeddedWallet as CrossAppEmbeddedWallet,
    type CrossAppSmartWallet as CrossAppSmartWallet,
    type LinkedAccountCrossApp as LinkedAccountCrossApp,
    type LinkedAccountAuthorizationKey as LinkedAccountAuthorizationKey,
    type LinkedAccountType as LinkedAccountType,
    type CustomMetadata as CustomMetadata,
    type LinkedAccountWalletInput as LinkedAccountWalletInput,
    type LinkedAccountEmailInput as LinkedAccountEmailInput,
    type LinkedAccountPhoneInput as LinkedAccountPhoneInput,
    type LinkedAccountGoogleInput as LinkedAccountGoogleInput,
    type LinkedAccountTwitterInput as LinkedAccountTwitterInput,
    type LinkedAccountDiscordInput as LinkedAccountDiscordInput,
    type LinkedAccountGitHubInput as LinkedAccountGitHubInput,
    type LinkedAccountSpotifyInput as LinkedAccountSpotifyInput,
    type LinkedAccountInstagramInput as LinkedAccountInstagramInput,
    type LinkedAccountTiktokInput as LinkedAccountTiktokInput,
    type LinkedAccountLineInput as LinkedAccountLineInput,
    type LinkedAccountTwitchInput as LinkedAccountTwitchInput,
    type LinkedAccountAppleInput as LinkedAccountAppleInput,
    type LinkedAccountLinkedInInput as LinkedAccountLinkedInInput,
    type LinkedAccountFarcasterInput as LinkedAccountFarcasterInput,
    type LinkedAccountTelegramInput as LinkedAccountTelegramInput,
    type LinkedAccountCustomJwtInput as LinkedAccountCustomJwtInput,
    type LinkedAccountInput as LinkedAccountInput,
    type UserBatchCreateInput as UserBatchCreateInput,
    type SMSMfaMethod as SMSMfaMethod,
    type TotpMfaMethod as TotpMfaMethod,
    type PasskeyMfaMethod as PasskeyMfaMethod,
    type LinkedMfaMethod as LinkedMfaMethod,
    type OAuthTokens as OAuthTokens,
    type UsersCursor as UsersCursor,
    type UserCreateParams as UserCreateParams,
    type UserListParams as UserListParams,
    type UserGetByCustomAuthIDParams as UserGetByCustomAuthIDParams,
    type UserGetByDiscordUsernameParams as UserGetByDiscordUsernameParams,
    type UserGetByEmailAddressParams as UserGetByEmailAddressParams,
    type UserGetByFarcasterIDParams as UserGetByFarcasterIDParams,
    type UserGetByGitHubUsernameParams as UserGetByGitHubUsernameParams,
    type UserGetByPhoneNumberParams as UserGetByPhoneNumberParams,
    type UserGetBySmartWalletAddressParams as UserGetBySmartWalletAddressParams,
    type UserGetByTelegramUserIDParams as UserGetByTelegramUserIDParams,
    type UserGetByTelegramUsernameParams as UserGetByTelegramUsernameParams,
    type UserGetByTwitterSubjectParams as UserGetByTwitterSubjectParams,
    type UserGetByTwitterUsernameParams as UserGetByTwitterUsernameParams,
    type UserGetByWalletAddressParams as UserGetByWalletAddressParams,
    type UserPregenerateWalletsParams as UserPregenerateWalletsParams,
    type UserSearchParams as UserSearchParams,
    type UserSetCustomMetadataParams as UserSetCustomMetadataParams,
    type UserUnlinkLinkedAccountParams as UserUnlinkLinkedAccountParams,
  };

  export { Analytics as Analytics, type AnalyticsEventInput as AnalyticsEventInput };

  export {
    Policies as Policies,
    type Policy as Policy,
    type SuiTransactionCommandOperator as SuiTransactionCommandOperator,
    type SuiTransferObjectsCommandField as SuiTransferObjectsCommandField,
    type TronTransactionCondition as TronTransactionCondition,
    type SuiTransactionCommandCondition as SuiTransactionCommandCondition,
    type SuiTransferObjectsCommandCondition as SuiTransferObjectsCommandCondition,
    type PolicyCreateRuleResponse as PolicyCreateRuleResponse,
    type PolicyDeleteResponse as PolicyDeleteResponse,
    type PolicyDeleteRuleResponse as PolicyDeleteRuleResponse,
    type PolicyUpdateRuleResponse as PolicyUpdateRuleResponse,
    type PolicyGetRuleResponse as PolicyGetRuleResponse,
    type PolicyCreateParams as PolicyCreateParams,
    type PolicyCreateRuleParams as PolicyCreateRuleParams,
    type PolicyDeleteParams as PolicyDeleteParams,
    type PolicyDeleteRuleParams as PolicyDeleteRuleParams,
    type PolicyUpdateParams as PolicyUpdateParams,
    type PolicyUpdateRuleParams as PolicyUpdateRuleParams,
    type PolicyGetRuleParams as PolicyGetRuleParams,
  };

  export {
    Transactions as Transactions,
    type TransactionScanningRequestBody as TransactionScanningRequestBody,
    type TransactionScanningResponseBody as TransactionScanningResponseBody,
    type TransactionGetResponse as TransactionGetResponse,
  };

  export {
    KeyQuorums as KeyQuorums,
    type KeyQuorum as KeyQuorum,
    type KeyQuorumDeleteResponse as KeyQuorumDeleteResponse,
    type KeyQuorumCreateParams as KeyQuorumCreateParams,
    type KeyQuorumDeleteParams as KeyQuorumDeleteParams,
    type KeyQuorumUpdateParams as KeyQuorumUpdateParams,
  };

  export {
    Aggregations as Aggregations,
    type AggregationMethod as AggregationMethod,
    type AggregationMetric as AggregationMetric,
    type AggregationWindow as AggregationWindow,
    type AggregationGroupBy as AggregationGroupBy,
    type Aggregation as Aggregation,
  };
}
