// @privy-io/api-types/resources/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Wallets = exports.Users = exports.Transactions = exports.Policies = exports.KeyQuorums = exports.Funding = exports.EmbeddedWallets = exports.CrossApp = exports.ClientAuth = exports.Apps = exports.Analytics = exports.Aggregations = void 0;
var aggregations_1 = require("./aggregations.js");
Object.defineProperty(exports, "Aggregations", { enumerable: true, get: function () { return aggregations_1.Aggregations; } });
var analytics_1 = require("./analytics.js");
Object.defineProperty(exports, "Analytics", { enumerable: true, get: function () { return analytics_1.Analytics; } });
var apps_1 = require("./apps.js");
Object.defineProperty(exports, "Apps", { enumerable: true, get: function () { return apps_1.Apps; } });
var client_auth_1 = require("./client-auth.js");
Object.defineProperty(exports, "ClientAuth", { enumerable: true, get: function () { return client_auth_1.ClientAuth; } });
var cross_app_1 = require("./cross-app.js");
Object.defineProperty(exports, "CrossApp", { enumerable: true, get: function () { return cross_app_1.CrossApp; } });
var embedded_wallets_1 = require("./embedded-wallets.js");
Object.defineProperty(exports, "EmbeddedWallets", { enumerable: true, get: function () { return embedded_wallets_1.EmbeddedWallets; } });
var funding_1 = require("./funding.js");
Object.defineProperty(exports, "Funding", { enumerable: true, get: function () { return funding_1.Funding; } });
var key_quorums_1 = require("./key-quorums.js");
Object.defineProperty(exports, "KeyQuorums", { enumerable: true, get: function () { return key_quorums_1.KeyQuorums; } });
var policies_1 = require("./policies.js");
Object.defineProperty(exports, "Policies", { enumerable: true, get: function () { return policies_1.Policies; } });
var transactions_1 = require("./transactions.js");
Object.defineProperty(exports, "Transactions", { enumerable: true, get: function () { return transactions_1.Transactions; } });
var users_1 = require("./users.js");
Object.defineProperty(exports, "Users", { enumerable: true, get: function () { return users_1.Users; } });
var wallets_1 = require("./wallets/wallets.js");
Object.defineProperty(exports, "Wallets", { enumerable: true, get: function () { return wallets_1.Wallets; } });
