// @privy-io/api-types/src/resources/embedded-wallets.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

import { APIResource } from '../core/resource';

export class EmbeddedWallets extends APIResource {}

/**
 * The request body for getting the iCloud recovery configuration.
 */
export interface RecoveryConfigurationICloudRequestBody {
  client_type: 'web' | 'expo-ios';
}

/**
 * The response body for getting the iCloud recovery configuration.
 */
export interface RecoveryConfigurationICloudResponseBody {
  api_token: string;

  container_identifier: string;

  environment: string;
}

/**
 * The request body for getting the recovery key material.
 */
export interface RecoveryKeyMaterialRequestBody {
  chain_type?: string;
}

/**
 * The response body for getting the recovery key material.
 */
export interface RecoveryKeyMaterialResponseBody {
  file_id: string;

  icloud_record_name: string;

  recovery_code: string;

  recovery_key: string;

  recovery_key_derivation_salt: string;

  recovery_type:
    | 'user_passcode_derived_recovery_key'
    | 'privy_passcode_derived_recovery_key'
    | 'privy_generated_recovery_key'
    | 'google_drive_recovery_secret'
    | 'icloud_recovery_secret';
}

/**
 * The response body for authenticating with OAuth for recovery.
 */
export interface RecoveryOAuthAuthenticateResponseBody {
  access_token: string;
}

/**
 * The request body for calling the OAuth callback for iCloud Expo.
 */
export interface RecoveryOAuthCallbackICloudExpoRequestBody {
  ckWebAuthToken: string;
}

/**
 * The request body for initiating an iCloud OAuth recovery flow.
 */
export interface RecoveryOAuthInitICloudRequestBody {
  client_type: 'web' | 'expo-ios';
}

/**
 * The request body for initiating an OAuth recovery flow.
 */
export interface RecoveryOAuthInitRequestBody {
  redirect_to: string;

  token?: string;

  code_challenge?: string;

  state_code?: string;
}

/**
 * The configuration object for smart wallets.
 */
export type SmartWalletConfiguration =
  | SmartWalletConfiguration.Enabled
  | SmartWalletConfiguration.UnionMember1;

export namespace SmartWalletConfiguration {
  export interface Enabled {
    enabled: false;
  }

  export interface UnionMember1 {
    configured_networks: Array<UnionMember1.ConfiguredNetwork>;

    enabled: true;

    smart_wallet_type:
      | 'safe'
      | 'kernel'
      | 'light_account'
      | 'biconomy'
      | 'coinbase_smart_wallet'
      | 'thirdweb';

    smart_wallet_version?: string;
  }

  export namespace UnionMember1 {
    export interface ConfiguredNetwork {
      bundler_url: string;

      chain_id: string;

      chain_name?: string;

      paymaster_context?: ConfiguredNetwork.PaymasterContext;

      paymaster_url?: string;

      rpc_url?: string;
    }

    export namespace ConfiguredNetwork {
      export interface PaymasterContext {
        policy_id: string;
      }
    }
  }
}

/**
 * The input configuration object for smart wallets.
 */
export type SmartWalletConfigurationInput =
  | SmartWalletConfigurationInput.Enabled
  | SmartWalletConfigurationInput.UnionMember1;

export namespace SmartWalletConfigurationInput {
  export interface Enabled {
    enabled: false;
  }

  export interface UnionMember1 {
    configured_networks: Array<UnionMember1.ConfiguredNetwork>;

    enabled: true;

    smart_wallet_type:
      | 'safe'
      | 'kernel'
      | 'light_account'
      | 'biconomy'
      | 'coinbase_smart_wallet'
      | 'thirdweb';

    smart_wallet_version?: string | null;
  }

  export namespace UnionMember1 {
    export interface ConfiguredNetwork {
      chain_id: string;

      bundler_url?: string;

      chain_name?: string;

      paymaster_context?: ConfiguredNetwork.PaymasterContext;

      paymaster_url?: string;

      rpc_url?: string;
    }

    export namespace ConfiguredNetwork {
      export interface PaymasterContext {
        policy_id: string;
      }
    }
  }
}

export declare namespace EmbeddedWallets {
  export {
    type RecoveryConfigurationICloudRequestBody as RecoveryConfigurationICloudRequestBody,
    type RecoveryConfigurationICloudResponseBody as RecoveryConfigurationICloudResponseBody,
    type RecoveryKeyMaterialRequestBody as RecoveryKeyMaterialRequestBody,
    type RecoveryKeyMaterialResponseBody as RecoveryKeyMaterialResponseBody,
    type RecoveryOAuthAuthenticateResponseBody as RecoveryOAuthAuthenticateResponseBody,
    type RecoveryOAuthCallbackICloudExpoRequestBody as RecoveryOAuthCallbackICloudExpoRequestBody,
    type RecoveryOAuthInitICloudRequestBody as RecoveryOAuthInitICloudRequestBody,
    type RecoveryOAuthInitRequestBody as RecoveryOAuthInitRequestBody,
    type SmartWalletConfiguration as SmartWalletConfiguration,
    type SmartWalletConfigurationInput as SmartWalletConfigurationInput,
  };
}
