// @privy-io/api-types/src/resources/funding.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

import { APIResource } from '../core/resource';

export class Funding extends APIResource {}

/**
 * The input for initializing a Coinbase on-ramp session.
 */
export type CoinbaseOnRampInitInput =
  | CoinbaseOnRampInitInput.UnionMember0
  | CoinbaseOnRampInitInput.UnionMember1;

export namespace CoinbaseOnRampInitInput {
  export interface UnionMember0 {
    addresses: Array<UnionMember0.Address>;

    assets?: Array<'eth' | 'ETH' | 'USDC' | 'POL'>;
  }

  export namespace UnionMember0 {
    export interface Address {
      address: string;

      blockchains: Array<
        | 'ethereum'
        | 'bitcoin'
        | 'base'
        | 'avacchain'
        | 'optimism'
        | 'solana'
        | 'polygon'
        | 'arbitrum'
        | 'stellar'
      >;
    }
  }

  export interface UnionMember1 {
    addresses: Array<UnionMember1.Address>;

    assets?: Array<'SOL' | 'USDC'>;
  }

  export namespace UnionMember1 {
    export interface Address {
      address: string;

      blockchains: Array<
        | 'ethereum'
        | 'bitcoin'
        | 'base'
        | 'avacchain'
        | 'optimism'
        | 'solana'
        | 'polygon'
        | 'arbitrum'
        | 'stellar'
      >;
    }
  }
}

/**
 * The response from initializing a Coinbase on-ramp session.
 */
export interface CoinbaseOnRampInitResponse {
  app_id: string;

  channel_id: string;

  partner_user_id: string;

  session_token: string;
}

/**
 * The response for getting the status of a Coinbase on-ramp session.
 */
export interface CoinbaseOnRampStatusResponse {
  status: 'pending' | 'success' | 'failure';
}

/**
 * The input for signing a Moonpay on-ramp.
 */
export type MoonpayOnRampSignInput =
  | MoonpayOnRampSignInput.UnionMember0
  | MoonpayOnRampSignInput.UnionMember1;

export namespace MoonpayOnRampSignInput {
  export interface UnionMember0 {
    address: string;

    config: UnionMember0.Config;

    useSandbox?: boolean;
  }

  export namespace UnionMember0 {
    export interface Config {
      currencyCode?:
        | 'AVAX_CCHAIN'
        | 'CELO_CELO'
        | 'CUSD_CELO'
        | 'DAI_ETHEREUM'
        | 'ETH_ETHEREUM'
        | 'ETH_ARBITRUM'
        | 'ETH_OPTIMISM'
        | 'ETH_POLYGON'
        | 'ETH_BASE'
        | 'FIL_FVM'
        | 'MATIC_ETHEREUM'
        | 'MATIC_POLYGON'
        | 'POL_POLYGON'
        | 'POL_ETHEREUM'
        | 'USDC_ETHEREUM'
        | 'USDC_ARBITRUM'
        | 'USDC_OPTIMISM'
        | 'USDC_POLYGON'
        | 'USDC_BASE'
        | 'USDC_CCHAIN'
        | 'USDC_SOL'
        | 'USDT_ETHEREUM'
        | 'USDT_POLYGON'
        | 'WETH_POLYGON'
        | 'WBTC_ETHEREUM'
        | 'BNB_BNB'
        | 'BNB_BSC'
        | 'CELO'
        | 'CUSD'
        | 'DAI'
        | 'ETH'
        | 'FIL'
        | 'MATIC'
        | 'USDC'
        | 'USDT'
        | 'WETH'
        | 'WBTC';

      email?: string;

      paymentMethod?:
        | 'ach_bank_transfer'
        | 'credit_debit_card'
        | 'gbp_bank_transfer'
        | 'gbp_open_banking_payment'
        | 'mobile_wallet'
        | 'sepa_bank_transfer'
        | 'sepa_open_banking_payment'
        | 'pix_instant_payment'
        | 'yellow_card_bank_transfer';

      quoteCurrencyAmount?: number;

      uiConfig?: Config.UiConfig;
    }

    export namespace Config {
      export interface UiConfig {
        accentColor?: string;

        theme?: 'light' | 'dark';
      }
    }
  }

  export interface UnionMember1 {
    address: string;

    config: UnionMember1.Config;

    useSandbox?: boolean;
  }

  export namespace UnionMember1 {
    export interface Config {
      currencyCode?: 'SOL' | 'USDC_SOL';

      email?: string;

      paymentMethod?:
        | 'ach_bank_transfer'
        | 'credit_debit_card'
        | 'gbp_bank_transfer'
        | 'gbp_open_banking_payment'
        | 'mobile_wallet'
        | 'sepa_bank_transfer'
        | 'sepa_open_banking_payment'
        | 'pix_instant_payment'
        | 'yellow_card_bank_transfer';

      quoteCurrencyAmount?: number;

      uiConfig?: Config.UiConfig;
    }

    export namespace Config {
      export interface UiConfig {
        accentColor?: string;

        theme?: 'light' | 'dark';
      }
    }
  }
}

/**
 * The response from signing a Moonpay on-ramp.
 */
export interface MoonpayOnRampSignResponse {
  externalTransactionId: string;

  signedUrl: string;
}

export declare namespace Funding {
  export {
    type CoinbaseOnRampInitInput as CoinbaseOnRampInitInput,
    type CoinbaseOnRampInitResponse as CoinbaseOnRampInitResponse,
    type CoinbaseOnRampStatusResponse as CoinbaseOnRampStatusResponse,
    type MoonpayOnRampSignInput as MoonpayOnRampSignInput,
    type MoonpayOnRampSignResponse as MoonpayOnRampSignResponse,
  };
}
