// @privy-io/api-types/src/resources/transactions.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

import { APIResource } from '../core/resource';
import { APIPromise } from '../core/api-promise';
import { RequestOptions } from '../internal/request-options';
import { path } from '../internal/utils/path';

export class Transactions extends APIResource {
  /**
   * Get a transaction by transaction ID.
   *
   * @example
   * ```ts
   * const transaction = await client.transactions.get(
   *   'transaction_id',
   * );
   * ```
   */
  get(transactionID: string, options?: RequestOptions): APIPromise<TransactionGetResponse> {
    return this._client.get(path`/v1/transactions/${transactionID}`, options);
  }
}

/**
 * The request body for scanning a transaction.
 */
export interface TransactionScanningRequestBody {
  chain_id: string;

  metadata: TransactionScanningRequestBody.Metadata;

  request: TransactionScanningRequestBody.Request;
}

export namespace TransactionScanningRequestBody {
  export interface Metadata {
    domain: string;
  }

  export interface Request {
    method: string;

    params: Array<unknown>;
  }
}

/**
 * The response from scanning a transaction.
 */
export interface TransactionScanningResponseBody {
  simulation: TransactionScanningResponseBody.UnionMember0 | TransactionScanningResponseBody.UnionMember1;

  validation: TransactionScanningResponseBody.UnionMember0 | TransactionScanningResponseBody.UnionMember1;
}

export namespace TransactionScanningResponseBody {
  export interface UnionMember0 {
    error: string;

    status: 'Error';
  }

  export interface UnionMember1 {
    assets_diffs: Array<UnionMember1.AssetsDiff>;

    exposures: Array<UnionMember1.Exposure>;

    status: 'Success';

    params?: UnionMember1.Params;
  }

  export namespace UnionMember1 {
    export interface AssetsDiff {
      asset: AssetsDiff.Asset;

      in: Array<AssetsDiff.In>;

      out: Array<AssetsDiff.Out>;
    }

    export namespace AssetsDiff {
      export interface Asset {
        decimals?: number;

        logo_url?: string;

        name?: string;

        symbol?: string;

        type?: string;
      }

      export interface In {
        usd_price?: string;

        value?: string;
      }

      export interface Out {
        usd_price?: string;

        value?: string;
      }
    }

    export interface Exposure {
      asset: Exposure.Asset;

      spenders: { [key: string]: Exposure.Spenders };
    }

    export namespace Exposure {
      export interface Asset {
        decimals?: number;

        logo_url?: string;

        name?: string;

        symbol?: string;

        type?: string;
      }

      export interface Spenders {
        usd_price?: string;

        value?: string;
      }
    }

    export interface Params {
      block_tag?: string;

      calldata?: Params.Calldata;

      chain?: string;

      data?: string;

      from?: string;

      gas?: string;

      gas_price?: string;

      to?: string;

      value?: string;
    }

    export namespace Params {
      export interface Calldata {
        function_selector: string;

        function_declaration?: string;

        function_signature?: string;
      }
    }
  }

  export interface UnionMember0 {
    error: string;

    status: 'Error';
  }

  export interface UnionMember1 {
    result_type: string;

    status: 'Success';
  }
}

export interface TransactionGetResponse {
  id: string;

  caip2: string;

  created_at: number;

  status:
    | 'broadcasted'
    | 'confirmed'
    | 'execution_reverted'
    | 'failed'
    | 'replaced'
    | 'finalized'
    | 'provider_error'
    | 'pending';

  transaction_hash: string | null;

  wallet_id: string;

  sponsored?: boolean;
}

export declare namespace Transactions {
  export {
    type TransactionScanningRequestBody as TransactionScanningRequestBody,
    type TransactionScanningResponseBody as TransactionScanningResponseBody,
    type TransactionGetResponse as TransactionGetResponse,
  };
}
